Plutarch/Api/V1/AssocMap.hs

```diff
@@ -47,3 +122,339 @@ instance
+pdelete :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (k :--> PMap k v :--> PMap k v)
+pdelete = rebuildAtKey # plam id
+
+-- | Rebuild the map at the given key.
+rebuildAtKey ::
+  (POrd k, PIsData k) =>
+  Term
+    s
+    ( ( PBuiltinList (PBuiltinPair (PAsData k) (PAsData v))
+          :--> PBuiltinList (PBuiltinPair (PAsData k) (PAsData v))
+      )
+        :--> k
+        :--> PMap k v
+        :--> PMap k v
+    )
+rebuildAtKey = phoistAcyclic $
+  plam $ \handler key map ->
+    punsafeFrom $
+      precList
+        ( \self x xs ->
+            plet (pfromData $ pfstBuiltin # x) $ \k ->
+              plam $ \prefix ->
+                pif
+                  (k #< key)
+                  (self # xs #$ plam $ \suffix -> prefix #$ pcons # x # suffix)
+                  ( pif
+                      (k #== key)
+                      (prefix #$ handler # xs)
+                      (prefix #$ handler #$ pcons # x # xs)
+                  )
+        )
+        (const $ plam (#$ handler # pnil))
+        # pto map
+        # plam id
+
+-- | Construct an empty 'PMap'.
+pempty :: Term (s :: S) (PMap k v)
+pempty = punsafeFrom pnil
+
+-- | Construct a singleton 'PMap' with the given key and value.
+psingleton :: (PIsData k, PIsData v) => Term (s :: S) (k :--> v :--> PMap k v)
+psingleton = phoistAcyclic $ plam $ \key value -> psingletonData # pdata key # pdata value
+
+-- | Construct a singleton 'PMap' with the given data-encoded key and value.
+psingletonData :: (PIsData k, PIsData v) => Term (s :: S) (PAsData k :--> PAsData v :--> PMap k v)
+psingletonData = phoistAcyclic $
+  plam $ \key value -> punsafeFrom (pcons # (ppairDataBuiltin # key # value) # pnil)
+
+-- | Construct a 'PMap' from a list of key-value pairs, sorted by ascending key data.
+pfromAscList :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (PBuiltinMap k v :--> PMap k v)
+pfromAscList = plam $ (passertSorted #) . pcon . PMap
+
+-- | Assert the map is properly sorted
+passertSorted :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (PMap k v :--> PMap k v)
+passertSorted = phoistAcyclic $
+  plam $ \map ->
+    precList
+      ( \self x xs ->
+          plet (pfromData $ pfstBuiltin # x) $ \k ->
+            plam $ \badKey ->
+              pif
+                (badKey # k)
+                (ptraceError "unsorted map")
+                (self # xs # plam (#< k))
+      )
+      (const $ plam $ const map)
+      # pto map
+      # plam (const $ pcon PFalse)
+
+data MapUnion k v f = MapUnion
+  { merge :: f (PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+  , mergeInsert :: f (PBuiltinPair (PAsData k) (PAsData v) :--> PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+  }
+
+type instance
+  ScottEncoded (MapUnion k v) a =
+    (PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+      :--> (PBuiltinPair (PAsData k) (PAsData v) :--> PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+      :--> a
+
+$(Rank2.TH.deriveAll ''MapUnion)
```

Comment:
Please no TH in Plutarch. TH is in general a PITA.

Reply:
You can do this in other ways, by e.g. having `(x : Bool) -> if x then A else B` as the "type" (you'd have to use `POpaque` since we don't have dependent types).

Reply:
I've inlined and simplified the instances.